File: cryptodevtest.pas
{* This simple test app is created with the sole purpose to check the validity
 * of the structures and ioctl IDs as defined in cryptodev.pas header *}

program cryptodevtest;

uses cryptodev;

const
  SEPARATOR = '=============================================================';

procedure DumpStructSizes;
begin
  writeln('Struct/Record sizes (in bytes):');
  writeln(SEPARATOR);
  writeln('TSessionOp:     ',sizeof(TSessionOp));
  writeln('TAlgInfo:       ',sizeof(TAlgInfo));
  writeln('TSessionInfoOp: ',sizeof(TSessionInfoOp));
  writeln('TCryptOp:       ',sizeof(TCryptOp));
  writeln('TCryptAuthOp:   ',sizeof(TCryptAuthOp));
  writeln('TCryptKOp:      ',sizeof(TCryptKOp));
end;

procedure DumpIOCtlIDs;
begin
  writeln('ioctl IDs:');
  writeln(SEPARATOR);
  writeln('CRIOGET          $',LowerCase(HexStr(CRIOGET,8)));
  writeln('CIOCGSESSION     $',LowerCase(HexStr(CIOCGSESSION,8)));
  writeln('CIOCFSESSION     $',LowerCase(HexStr(CIOCFSESSION,8)));
  writeln('CIOCCRYPT        $',LowerCase(HexStr(CIOCCRYPT,8)));
  writeln('CIOCKEY          $',LowerCase(HexStr(CIOCKEY,8)));
  writeln('CIOCASYMFEAT     $',LowerCase(HexStr(CIOCASYMFEAT,8)));
  writeln('CIOCGSESSIONINFO $',LowerCase(HexStr(CIOCGSESSINFO,8)));
  writeln('CIOCAUTHCRYPT    $',LowerCase(HexStr(CIOCAUTHCRYPT,8)));
  writeln('CIOCASYNCCRYPT   $',LowerCase(HexStr(CIOCASYNCCRYPT,8)));
  writeln('CIOCASYNCFETCH   $',LowerCase(HexStr(CIOCASYNCFETCH,8)));
end;

begin
  writeln;
  DumpStructSizes;
  writeln;
  DumpIOCtlIDs;
  writeln;
end.

File: cryptodevtest.c
/* This simple test app is created with the sole purpose to check the validity
 * of the structures and ioctl IDs as defined in cryptodev.h against the 
 * cryptodev.pas Pascal header */

#include <stdio.h>

#include <linux/ioctl.h>
#include "crypto/cryptodev.h"

#define SEPARATOR "=============================================================\n"

void dump_struct_sizes()
{
  printf("Struct/Record sizes (in bytes):\n");
  printf(SEPARATOR);
  printf("TSessionOp:     %d\n",sizeof(struct session_op));
  printf("TAlgInfo:       %d\n",sizeof(struct alg_info));
  printf("TSessionInfoOp: %d\n",sizeof(struct session_info_op));
  printf("TCryptOp:       %d\n",sizeof(struct crypt_op));
  printf("TCryptAuthOp:   %d\n",sizeof(struct crypt_auth_op));
  printf("TCryptKOp:      %d\n",sizeof(struct crypt_kop));
}

void dump_ioctl_ids()
{
  printf("ioctl IDs:\n");
  printf(SEPARATOR);
  printf("CRIOGET          $%x\n",CRIOGET);
  printf("CIOCGSESSION     $%x\n",CIOCGSESSION);
  printf("CIOCFSESSION     $%x\n",CIOCFSESSION);
  printf("CIOCCRYPT        $%x\n",CIOCCRYPT);
  printf("CIOCKEY          $%x\n",CIOCKEY);
  printf("CIOCASYMFEAT     $%x\n",CIOCASYMFEAT);
  printf("CIOCGSESSIONINFO $%x\n",CIOCGSESSINFO);
  printf("CIOCAUTHCRYPT    $%x\n",CIOCAUTHCRYPT);
  printf("CIOCASYNCCRYPT   $%x\n",CIOCASYNCCRYPT);
  printf("CIOCASYNCFETCH   $%x\n",CIOCASYNCFETCH);
};

int main(int argc, char** argv)
{
  printf("\n");
  dump_struct_sizes();
  printf("\n");
  dump_ioctl_ids();
  printf("\n");

  return 0;
}
